### DISCUSSION TITLE
为什么Xray能直接根据domain进行路由，不是先有DNS请求才有web request？

### DISCUSSION THREAD

**Discussion by tlsalex** [XTLS/Xray-core#1113] [General]
如题 -- 为什么Xray能直接根据domain进行路由，不是先有DNS请求才有web request吗？DNS请求得到IP，Xray是不是应该根据IP进行路由？

在客户端里，XRay能先看到DNS请求还是先看到web request？

我对web request 不是很理解，我想用以下两种形式来表达我的理解，
A:
1. 在浏览器里输入www.oversea.com
2. 浏览器需要先把www.oversea.com 解析成IP 地址，这时浏览器会不会是 这样问Xray客户端,： 亲，你能帮我解析这个域名吗？
3. Xray客户端 看到是oversea 的网站，并尝试解析，但这时Xray怎么判断用哪个DNS去解析呢，是根据路由配置，oversea 的要发去Xray 服务器？然后把请求发给Xray服务器： 哥们，能否帮我解析这个域名？
4. Xray服务器看到DNS解析请求，就用Xray所配置的DNS去解析这个域名，并把结果返回Xray客户端。
5. Xray客户端把IP返回给浏览器，浏览器用这个IP发出web request. 
6. 
```
Request URL: https://www.oversea.com/
Request Method: GET
Status Code: 200 OK
Remote Address: 127.0.0.1:1081
Referrer Policy: strict-origin-when-cross-origin
```
7. Xray 客户端收到web request , request 里应该是包含了 域名和IP --（我猜测的，我不懂），Xray 客户端这时就可以根据域名或IP来进行路由。

Q: 一个web请求 是不是出现了两个路由过程，一个是把解析DNS请求路由到合适的出口，另外一个是把web request 路由到合适的出口？

B: 
1. 在浏览器里输入www.oversea.com
2. Xray客户端能立马看到浏览器要访问是哪个域名？
3. Xray客户端即可根据域名路由到合适的出口。但这怎么解释DNS部分呢？如何在xray里验证能直接看到web request而不是先看到DNS请求。

可能是我对web 相关的过程不理解， 各位可否帮忙科普下呢

**Comment by BioniCosmos** [comment]
如果使用 Socks 5 协议不会进行 DNS 请求，直接发送域名。
如果使用透明代理，sniffing 会嗅探出域名。

**Comment by azzvx** [comment]
对于显式通过http或者socks代理访问Xray，请求的域名是直接发送给Xray的。
对于通过透明代理访问Xray，入站协议的sniffing模块能检测连接请求的域名，因为不论http、tls还是quic，它请求的域名都是明文显示的。

**Comment by chika0801** [comment]
https://tachyondevel.medium.com/%E6%BC%AB%E8%B0%88%E5%90%84%E7%A7%8D%E9%BB%91%E7%A7%91%E6%8A%80%E5%BC%8F-dns-%E6%8A%80%E6%9C%AF%E5%9C%A8%E4%BB%A3%E7%90%86%E7%8E%AF%E5%A2%83%E4%B8%AD%E7%9A%84%E5%BA%94%E7%94%A8-62c50e58cbd0
这文章能帮到你，这些知识几句话说不清楚，我也是看这文章自己慢慢懂的
